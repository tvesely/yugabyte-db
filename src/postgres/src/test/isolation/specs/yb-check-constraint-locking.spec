setup
{
  CREATE TABLE products (
    product_no integer PRIMARY KEY,
    name text,
    price numeric,
    CHECK (price > 0),
    discounted_price numeric,
    CHECK (discounted_price > 0),
    CHECK (price > discounted_price)
  );

  INSERT INTO products VALUES (1, 'oats', 10, 1);
}

teardown
{
  DROP TABLE products;
}

session "s1"
setup		{ BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED; }
step "s1a"	{ UPDATE PRODUCTS SET price = 3 WHERE product_no = 1; }
step "s1b"	{ SELECT price, discounted_price FROM products; }
step "s1c"	{ COMMIT; }


session "s2"
step "s2s_start"  { BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE; }
step "s2rr_start" { BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; }
step "s2rc_start" { BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED; }
step "s2a"	{ UPDATE PRODUCTS SET discounted_price = 4 WHERE product_no = 1; }
step "s2b"	{ COMMIT; }

# This test is to ensure that the CHECK constraint is taken into account by concurrent transactions
permutation "s2s_start" "s1a" "s2a" "s1b" "s1c" "s2b" "s1b"
permutation "s2rr_start" "s1a" "s2a" "s1b" "s1c" "s2b" "s1b"
permutation "s2rc_start" "s1a" "s2a" "s1b" "s1c" "s2b" "s1b"
