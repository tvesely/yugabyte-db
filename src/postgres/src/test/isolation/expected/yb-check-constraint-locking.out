Parsed test spec with 2 sessions

starting permutation: s2s_start s1a s2a s1b s1c s2b s1b
step s2s_start: BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;
step s1a: UPDATE PRODUCTS SET price = 3 WHERE product_no = 1;
step s2a: UPDATE PRODUCTS SET discounted_price = 4 WHERE product_no = 1; <waiting ...>
step s1b: SELECT price, discounted_price FROM products;
price          discounted_price

3              1              
step s1c: COMMIT;
step s2a: <... completed>
error in steps s1c s2a: ERROR:  could not serialize access due to concurrent update
step s2b: COMMIT;
step s1b: SELECT price, discounted_price FROM products;
price          discounted_price

3              1              

starting permutation: s2rr_start s1a s2a s1b s1c s2b s1b
step s2rr_start: BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ;
step s1a: UPDATE PRODUCTS SET price = 3 WHERE product_no = 1;
step s2a: UPDATE PRODUCTS SET discounted_price = 4 WHERE product_no = 1; <waiting ...>
step s1b: SELECT price, discounted_price FROM products;
price          discounted_price

3              1              
step s1c: COMMIT;
step s2a: <... completed>
error in steps s1c s2a: ERROR:  could not serialize access due to concurrent update
step s2b: COMMIT;
step s1b: SELECT price, discounted_price FROM products;
price          discounted_price

3              1              

starting permutation: s2rc_start s1a s2a s1b s1c s2b s1b
step s2rc_start: BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
step s1a: UPDATE PRODUCTS SET price = 3 WHERE product_no = 1;
step s2a: UPDATE PRODUCTS SET discounted_price = 4 WHERE product_no = 1; <waiting ...>
step s1b: SELECT price, discounted_price FROM products;
price          discounted_price

3              1              
step s1c: COMMIT;
step s2a: <... completed>
error in steps s1c s2a: ERROR:  new row for relation "products" violates check constraint "products_check"
step s2b: COMMIT;
step s1b: SELECT price, discounted_price FROM products;
price          discounted_price

3              1              
